src/esp/assets/ResourceManager.cpp

```diff
@@ -2184,5 +2188,188 @@ std::unique_ptr<MeshData> ResourceManager::createJoinedCollisionMesh(
+      Mn::Utility::formatInto(out, out.size(), "{0} {1} {2} {3}{4}", "v",
+                              pos[0], pos[1], pos[2], "\n");
+      numVertices++;
+    }
+  }
+
+  Mn::Utility::formatInto(out, out.size(), "{0} {1} {2}",
+                          "# Number of vertices", numVertices, "\n\n");
+
+  // Now do second pass to write indices for each group (node)
+  int globalVertexNum = 1;
+  int numParts = 1;
+  for (const MeshTransformNode& node : metaData.root.children) {
+    CollisionMeshData& meshData =
+        meshes_.at(node.meshIDLocal + metaData.meshIndex.first)
+            ->getCollisionMeshData();
+    Mn::Utility::formatInto(out, out.size(), "{0}{1} {2}", "g part_", numParts,
+                            "mesh\n");
+    for (int ix = 0; ix < meshData.indices.size(); ix += 3) {
+      Mn::Utility::formatInto(out, out.size(), "{0} {1} {2} {3}{4}", "f",
+                              meshData.indices[ix] + globalVertexNum,
+                              meshData.indices[ix + 1] + globalVertexNum,
+                              meshData.indices[ix + 2] + globalVertexNum, "\n");
+    }
+    numParts++;
+    globalVertexNum += meshData.positions.size();
+  }
+  Cr::Utility::Directory::appendString(filepath + "/" + new_filename, out);
+
+  return success;
+}
+
+unsigned int ResourceManager::getNumberOfResource(
+    const std::string& resourceName) {
+  return resourceDict_.count(resourceName);
+}
+
+#ifdef ESP_BUILD_WITH_VHACD
+void ResourceManager::createConvexHullDecomposition(
+    const std::string& filename,
+    const std::string& CHDFilename,
+    const VHACDParameters& params,
+    const bool saveCHDToObj) {
+  if (resourceDict_.count(filename) == 0) {
+    // load/check for render MeshMetaData and load assets
+    loadObjectMeshDataFromFile(filename, filename, "render", true);
+
+  }  // if no render asset exists
+
+  // get joined mesh data
+  assets::MeshData::uptr joinedMesh = assets::MeshData::create_unique();
+  joinedMesh = createJoinedCollisionMesh(filename);
+
+  // use VHACD
+  interfaceVHACD->Compute(&joinedMesh->vbo[0][0], joinedMesh->vbo.size(),
+                          &joinedMesh->ibo[0], joinedMesh->ibo.size() / 3,
+                          params);
+
+  Cr::Utility::Debug() << "== VHACD ran ==";
+
+  // convert convex hulls into MeshDatas, CollisionMeshDatas
+  int meshStart = meshes_.size();
+  std::vector<CollisionMeshData> collisionMeshGroup;
+  int nConvexHulls = interfaceVHACD->GetNConvexHulls();
+  Cr::Utility::Debug() << "Num Convex Hulls: " << nConvexHulls;
+  Cr::Utility::Debug() << "Resolution: " << params.m_resolution;
+  VHACD::IVHACD::ConvexHull ch{};
+  std::unique_ptr<GenericMeshData> genCHMeshData;
+  for (unsigned int p = 0; p < nConvexHulls; ++p) {
+    // for each convex hull, transfer the data to a newly created  MeshData
+    interfaceVHACD->GetConvexHull(p, ch);
+
+    std::vector<Magnum::UnsignedInt> indices;
+    std::vector<Magnum::Vector3> positions;
```

Comment:
[From my previous comment](https://github.com/facebookresearch/habitat-sim/pull/1055#discussion_r573129296) -- this doesn't seem to be done yet:

> Same below, unless VHACD is *really badly designed* (which I don't think it is), magnum's MeshData should be able to reference the data from `ch` directly, without having to copy again.